calyx/src/passes/top_down_static_timing.rs

```diff
@@ -78,33 +184,122 @@ fn seq_calculate_states(
     pre_guard: &ir::Guard,
     schedule: &mut Schedule,
     builder: &mut ir::Builder,
-) -> u64 {
-    let mut cur = cur_state;
+) -> CalyxResult<Vec<PredEdge>> {
+    get_static_time(&con.attributes);
+
+    let mut preds = vec![];
+    let default_pred = (cur_state, pre_guard.clone());
     for stmt in &con.stmts {
-        cur = calculate_states(stmt, cur, pre_guard, schedule, builder)
+        // Add transition(s) from last state to the new state.
+        let new_state = seq_add_transitions(schedule, &preds, &default_pred);
+
+        // Recurse into statement and save new predecessors.
+        match calculate_states(stmt, new_state, pre_guard, schedule, builder) {
+            Ok(inner_preds) => {
+                preds = inner_preds;
+            }
+            Err(e) => return Err(e),
+        }
     }
-    cur
+
+    // Add final transition(s) from the last statement.
+    seq_add_transitions(schedule, &preds, &default_pred);
```

Comment:
This got a little weird, and I haven't come back to see if there is a smarter way to address what I saw. The current algorithm seems to work well, but it wouldn't handle the final statement in a seq, so I added this here to address that. But this ended up making duplicate transitions in some cases (really, the only transition this line is needed to add is the final statement in the outermost seq, from what I saw). Perhaps there is a better way to address that case; what I chose was to make the transitions a set, rather than a vector, so any duplicates would be ignored.